File: Libraries/Services/PasswordGeneratorService.cs
using Models.DTOs;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace Services
{
    public static class PasswordGeneratorService
    {
        private const string _letters = "abcdefghijklmnopqrstuvwxyz";
        private const string _numbers = "0123456789";
        private const string _specialChars = "!@#$%^&*";

        public static string GeneratePassword(PasswordGeneratorSettingsDTO settings)
        {
            return PasswordGeneratorService.GeneratePassword(settings.PasswordLength, settings.UseNumbers, settings.UseSpecialChars, settings.UseUppercase, settings.UseLowercase);
        }

        public static string GeneratePassword(int length = 8, bool useNumbers = true, bool useSpecialChars = true, bool useUppercase = true, bool useLowercase = true)
        {
            if (length < 1)
                return string.Empty;

            StringBuilder sb = new StringBuilder();

            if (useNumbers) sb.Append(_numbers);
            if (useSpecialChars) sb.Append(_specialChars);
            if (useUppercase) sb.Append(_letters.ToUpperInvariant());
            if (useLowercase) sb.Append(_letters);

            string alphabet = sb.ToString();
            if (alphabet.Length == 0)
                return string.Empty;

            string password;
            do
            {
                sb.Clear();
                for (int i = 0; i < length; i++)
                {
                    int next = RandomNumberGenerator.GetInt32(alphabet.Length);
                    sb.Append(alphabet[next]);
                }
                password = sb.ToString();
            } while (!CheckConstraints(password, useNumbers, useSpecialChars, useUppercase, useLowercase));


            return password;
        }

        private static bool CheckConstraints(string text, PasswordGeneratorSettingsDTO settings)
        {
            return PasswordGeneratorService.CheckConstraints(text, settings.UseNumbers, settings.UseSpecialChars, settings.UseUppercase, settings.UseLowercase);
        }

        private static bool CheckConstraints(string text, bool useNumbers = true, bool useSpecialChars = true, bool useUppercase = true, bool useLowercase = true)
        {
            if (useNumbers && !text.Any(char.IsDigit))
                return false;
            else if (!useNumbers && text.Any(char.IsDigit))
                return false;

            if (useUppercase && !text.Any(char.IsUpper))
                return false;
            else if (!useUppercase && text.Any(char.IsUpper))
                return false;

            if (useLowercase && !text.Any(char.IsLower))
                return false;
            else if (!useLowercase && text.Any(char.IsLower))
                return false;

            string strRegex = @"[!@#$%^&*]";

            Regex re = new Regex(strRegex);

            if (useSpecialChars)
            {
                if (!re.IsMatch(text))
                    return false;
            }
            else if (!useSpecialChars)
            {
                if (re.IsMatch(text))
                    return false;
            }

            return true;
        }
    }
}

File: PasswordManager.AngularApp/src/app/services/password-generator.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from 'src/environments/environment';
import { AbstractControl, FormArray } from '@angular/forms';
import { Observable } from 'rxjs';
import { PasswordGeneratorSettingsResponse } from '../models/password-generator-settings-response.model';
import { ActivatedRoute } from '@angular/router';
import { PasswordGeneratorSettings } from '../models/password-generator-settings.model';

@Injectable({
  providedIn: 'root'
})
export class PasswordGeneratorService {

  private passwordsGeneratorPath = `${environment.apiUrl}/passwords/generator`;

  private readonly letters: string = "abcdefghijklmnopqrstuvwxyz";
  private readonly numbers: string = "0123456789";
  private readonly specialChars: string = "!@#$%^&*";

  constructor(private http: HttpClient, private route: ActivatedRoute) { }

  generatePassword(data: AbstractControl<any, any>): Observable<PasswordGeneratorSettingsResponse> {
    return this.http.post<PasswordGeneratorSettingsResponse>(this.passwordsGeneratorPath, data);
  }

  public generatePasswordFromModel(settings: PasswordGeneratorSettings): string {
    return this.generatePasswordWithOptions(settings.passwordLength, settings.useNumbers, settings.useSpecialChars, settings.useUppercase, settings.useLowercase);
  }

  public generatePasswordWithOptions(length: number = 8, useNumbers: boolean = true, useSpecialChars: boolean = true, useUppercase: boolean = true, useLowercase: boolean = true): string {
    if (length < 1) {
      return "";
    }

    let sb: string = "";

    if (useNumbers) {
      sb += this.numbers;
    }

    if (useSpecialChars) {
      sb += this.specialChars;
    }

    if (useUppercase) {
      sb += this.letters.toUpperCase();
    }

    if (useLowercase) {
      sb += this.letters;
    }

    const alphabet: string = sb.toString();
    if (alphabet.length === 0) {
      return "";
    }

    let password: string;
    do {
      sb = "";
      for (let i = 0; i < length; i++) {
        const next: number = this.getRandomInt(alphabet.length);
        sb += alphabet[next];
      }
      password = sb.toString();
    } while (!this.checkConstraints(password, useNumbers, useSpecialChars, useUppercase, useLowercase));

    return password;
  }

  private checkConstraints(text: string, useNumbers: boolean = true, useSpecialChars: boolean = true, useUppercase: boolean = true, useLowercase: boolean = true): boolean {
    if (useNumbers && !/\d/.test(text)) {
      return false;
    } else if (!useNumbers && /\d/.test(text)) {
      return false;
    }

    if (useUppercase && !/[A-Z]/.test(text)) {
      return false;
    } else if (!useUppercase && /[A-Z]/.test(text)) {
      return false;
    }

    if (useLowercase && !/[a-z]/.test(text)) {
      return false;
    } else if (!useLowercase && /[a-z]/.test(text)) {
      return false;
    }

    if (useSpecialChars) {
      if (!/[!@#$%^&*]/.test(text)) {
        return false;
      }
    } else if (!useSpecialChars) {
      if (/[!@#$%^&*]/.test(text)) {
        return false;
      }
    }

    return true;
  }

  private getRandomInt(max: number): number {
    const randomBytes: Uint8Array = new Uint8Array(1);
    window.crypto.getRandomValues(randomBytes);
    return randomBytes[0] % max;
  }

}
